openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -768,6 +768,59 @@ def _get_encrypted_request_headers(request):
         return encrypt_for_log(str(request.headers), getattr(settings, 'SAFE_SESSIONS_DEBUG_PUBLIC_KEY', None))
 
 
+class EmailChangeSessionInvalidationMiddleware(MiddlewareMixin):
+    """
+    Middleware responsible for invalidating user sessions
+    by detecting a mismatch between the user's email in
+    the session and request.user.email.
+
+    This middleware ensures that the sessions in other browsers
+    are invalidated when a user changes their email in one browser.
+    The active session in which the email change is made will remain valid.
+
+    The user's email is stored in their session during login
+    and gets updated when the user changes their email.
+    This middleware checks for any mismatch between the stored email
+    and the current user's email in each request, and if found,
+    it invalidates/flushes the session and mark cookies for deletion in request
+    which are then deleted in the process_response of SafeSessionMiddleware.
+    """
+
+    def process_request(self, request):
+        """
+        Invalidate the user session if there's a mismatch
+        between the email in the user's session and request.user.email.
+        """
+        if settings.ENFORCE_SESSION_EMAIL_MATCH and request.user.is_authenticated:
+            user_session_email = request.session.get('email', None)
+
+            if user_session_email is not None and request.user.email != user_session_email:
+                # Flush the session and mark cookies for deletion.
+                request.session.flush()
+                request.user = AnonymousUser()
+                _mark_cookie_for_deletion(request)
```

Comment:
A different version of this might be to set variable about whether or not we've matched, and have one custom attribute for whether or not we've matched, and another for the setting `ENFORCE_SESSION_EMAIL_MATCH`. Then flush if there is a mismatch and we are enforcing. In that way, you can observe before you enforce how many requests would be affected, in case there is an issue before you start enforcing. Something to consider.

Reply:
This is close. To complete what I meant (if you wish), you could rename `_set_session_email_match_custom_attribute` to `_set_session_email_match_custom_attributes`, and inside that method, you could add:
```
set_custom_attribute('is_enforce_session_email_match_enabled', settings.ENFORCE_SESSION_EMAIL_MATCH)
```
From in New Relic, this would just help you see which transactions were from before and after this was enabled.